opencog/openpsi/main.scm

```diff
@@ -134,15 +140,36 @@
 )
 
 ; -------------------------------------------------------------
-(define (psi-halt component)
+(define (save-thread thread component)
+ "
+  save-thread THREAD COMPONENT
+
+  Save thread in the psi-threads list to properly terminate it
+  before exit.
+ "
+ (set! psi-threads (cons (cons thread component) psi-threads))
+)
+
+; -------------------------------------------------------------
+(define* (psi-halt component #:optional (timeout 3))
 "
   psi-halt COMPONENT
 
   Halts COMPONENT's previously-started psi loop thread. The thread is
   started by calling `(psi-run COMPONENT)`.
 "
+  (define (psi-join lst)
+    (if (not (null? lst))
+      (let
+        ((thread (car (car lst)))
+         (comp (cdr (car lst))))
```

Comment:
So, `(car (car lst))` is the same as `(caar lst)` and `(cdr (car lst))` is the same as `(cdar lst)`  and there's a large collection of utilities here: https://srfi.schemers.org/srfi-1/srfi-1.html

Reply:
... and for accessing elements of a pair, you can use `first` and `second` .. or maybe even `list-ref`  ... here, it does not really matter, but names like that are just slightly more readable for conventional programmers coming from other languages.